### PR TITLE
Initial support for Qwen3.5 family

### PR DESCRIPTION
**Pull request by haok1402**
Adds training support for the Qwen3.5-MoE architecture, starting with Qwen3.5-35B-A3B. This is an initial cut: it covers the text tower only, and context parallelism is not yet implemented.

## What this adds

- **Model** (`pithtrain/models/qwen35_moe.py`): the Qwen3.5-MoE text model, implemented against the 5-stage DualPipeV layer protocol. It is a hybrid MoE: most layers use Gated DeltaNet (linear attention) and the rest use full softmax attention (output-gated GQA with partial rotary). Every layer is MoE, with routed experts plus a sigmoid-gated shared expert.
- **Operator** (`pithtrain/operators/gated_delta_rule.py`): wraps FLA's fused chunked gated delta rule as a `torch.library` custom op, so the linear-attention layers stay fullgraph torch.compile friendly. Adds `flash-linear-attention` as a dependency.
- **Checkpoint conversion** (`pithtrain/tasks/convert_checkpoint/qwen35_moe.py`): converts the released checkpoint to and from the runtime DCP format. The released model is vision-language, so only the text tower (`model.language_model.*` and `lm_head`) is converted; the vision tower and MTP weights are dropped. GPT-OSS detection is tightened to its per-expert bias key so the two fused-expert formats are not confused.
- **Examples and benchmarks**: pretrain, tokenize, and convert configs under `examples/.../qwen3.5-35b-a3b`, plus a pretraining setup with 2-node and 4-node configs under `benchmarks/pretraining/qwen3.5-35b-a3b`.
- **Tests** (`tests/test_fsdp.py`): extends the FSDP correctness ladder to the new model.

## Validation

- Single-GPU reference forward matches the 5-stage DualPipeV path.
- FSDP correctness ladder passes from pp1/ep1 up to pp2/ep2, with matching gradients.
- Checkpoint conversion round-trips byte for byte.
- End-to-end pretraining converges at pp2/ep8 across two nodes.

## Loaded-checkpoint correctness

Loading the converted release checkpoint and running a short job (pp2/ep8, sequence length 1536) starts at a cross-entropy loss around 2.4 on the very first step and holds there, rather than the high loss a randomly initialized model would show. This confirms the weight conversion and the forward path are correct end to end.

```
[rank0] layer_partition: 40 layers / 4 stages -> [10, 10, 10, 10]
[rank0] layer_partition: 40 layers / 4 stages -> [10, 10, 10, 10]
2026-06-22 18:28:45 | INFO | launch(cfg=PretrainLMCfg(distributed=DistributedCfg(pipeline_parallel_size=2, context_parallel_size=1, expert_parallel_size=8, timeout=datetime.timedelta(seconds=900), sharding_strategy='fsdp'), training=TrainingCfg(dataset=PosixPath('workspace/datasets/dclm-baseline/toktxt/qwen3.5'), sequence_length=1536, seed=1234, lr=1e-06, max_steps=25, micro_batch_size=1, global_batch_size=1024, optimizer=<function make_muon_optimizer at 0x134b422c8720>, scheduler=<function make_constant_scheduler at 0x134b422c8b40>, model=PosixPath('benchmarks/pretraining/qwen3.5-35b-a3b/model.json'), save_interval=None, save_location=PosixPath('workspace/checkpoints/qwen3.5-35b-a3b'), moe_load_balance_coef=0.001, moe_load_balance_type='global-batch', fp8_training='disabled', init_std=0.02, nsys_start=None, nsys_stop=None, memory_profile_start=None, memory_profile_stop=None, memory_profile_output=PosixPath('/home/haok/pith-train')), logging=LoggingCfg(wandb=None)))
2026-06-22 18:28:45 | INFO | Load checkpoint: workspace/checkpoints/qwen3.5-35b-a3b/torch-dcp/step-00000000
2026-06-22 18:28:52 | INFO | Load checkpoint: Elapsed min=7.4s, max=7.4s
2026-06-22 18:30:04 | INFO | step 00000001/00000025 | step-time 71.522 sec | cross-entropy-loss 2.4616 | load-balance-loss 1.066807 | learning-rate 1.000000e-06 | gradient-norm 355.1682 | tokens-per-second 21,991 | peak-gpu-memory 54.40 GB
2026-06-22 18:30:27 | INFO | step 00000002/00000025 | step-time 22.658 sec | cross-entropy-loss 2.4768 | load-balance-loss 1.064240 | learning-rate 1.000000e-06 | gradient-norm 373.1439 | tokens-per-second 69,417 | peak-gpu-memory 54.37 GB
2026-06-22 18:30:50 | INFO | step 00000003/00000025 | step-time 22.530 sec | cross-entropy-loss 2.4182 | load-balance-loss 1.072402 | learning-rate 1.000000e-06 | gradient-norm 360.1254 | tokens-per-second 69,812 | peak-gpu-memory 54.39 GB
2026-06-22 18:31:13 | INFO | step 00000004/00000025 | step-time 22.651 sec | cross-entropy-loss 2.3657 | load-balance-loss 1.066703 | learning-rate 1.000000e-06 | gradient-norm 337.9038 | tokens-per-second 69,440 | peak-gpu-memory 54.37 GB
2026-06-22 18:31:36 | INFO | step 00000005/00000025 | step-time 22.569 sec | cross-entropy-loss 2.4554 | load-balance-loss 1.070877 | learning-rate 1.000000e-06 | gradient-norm 268.8993 | tokens-per-second 69,692 | peak-gpu-memory 54.43 GB
```

### CODE REVIEW

**Comment by haok1402** [comment]
@claude are the changes compact? are there any abstractions introduced where you find helpful to avoid? are there any alternative implementations which may present a cleaner implementation? are there any legacy documents as a result of this change?

**Comment by haok1402** [comment]
> benchmarks/pretraining/qwen3.5-35b-a3b/model.json is a byte-for-byte duplicate of examples/pretrain_lm/qwen3.5-35b-a3b/config.json (diff produces no output). One of them is redundant. setup.py references benchmarks/.../model.json and examples/pretrain_lm/.../script.py references examples/.../config.json, so both paths are load-bearing — but the content is identical. The benchmark setup.py could simply point at the examples config (examples/pretrain_lm/qwen3.5-35b-a3b/config.json) to eliminate the duplicate.

Fixed. Benchmarking should be carried out for all models yet only a few models are demonstrated via examples. Therefore, we prefer re-using the `model.json` from the benchmarking folder instead.

**Comment by haok1402** [comment]
> **`Qwen35MoeRMSNormGated.forward`** (lines 80–86) has three dtype round-trips: cast to float32 → normalize → cast back to `input_dtype` → multiply by weight (implicit upcast to float32 because weight is float32) → `silu(gate.float())` → cast back to `input_dtype`. The Gemini suggestion is correct — do all computation in float32, cast once at the end:
> 
> ```python
> def forward(self, x: torch.Tensor, gate: torch.Tensor) -> torch.Tensor:
>     input_dtype = x.dtype
>     x_f = x.float()
>     x_norm = x_f * torch.rsqrt(x_f.pow(2).mean(-1, keepdim=True) + self.eps)
>     return (x_norm * self.weight.float() * F.silu(gate.float())).to(input_dtype)
> ```

Fixed.

**Comment by haok1402** [comment]
> **`Qwen35MoeRMSNorm.forward`** (line 63) calls `x.float()` twice:
> 
> ```python
> output = x.float() * torch.rsqrt(x.float().pow(2).mean(-1, keepdim=True) + self.eps)
> ```
> 
> The second `x.float()` is a redundant cast. The already-confirmed Gemini suggestion applies here:
> 
> ```python
> x_f = x.float()
> output = x_f * torch.rsqrt(x_f.pow(2).mean(-1, keepdim=True) + self.eps)
> output = output * (1.0 + self.weight.float())
> return output.type_as(x)
> ```

Fixed. Arguably speaking, the `torch.compile` should be able to optimize this away.

**Comment by haok1402** [comment]
> `Qwen35MoeTopKRouter` (lines 354–386) is `Qwen3MoeGate` with two features removed (`router_replay`, `norm_topk_prob` parameter). The compute logic is otherwise identical. Since `norm_topk_prob` is always `True` in Qwen3 too, the classes could be unified — or `Qwen35MoeTopKRouter` could simply reuse `Qwen3MoeGate` with default arguments.

We would by design prefer a per-model definition for the clarity of things in the sacrifice of some "perhaps-repetitive" code instead of many different configs that match-case around. However, that `router-replay` argument makes sense and I think the benchmarking file requires a cleanup too! thanks for catching this! Fixed.

**Comment by haok1402** [comment]
> **`Qwen35MoeExperts` reimplements the FP8 weight cache pattern** already present in `gpt_oss.py`. `_wq_cache`, `_wq_version`, `_quantized_weight()`, and `_group_linear()` are a second copy of the same caching logic. Meanwhile, `Qwen3MoeExperts` uses `get_group_linear_cls()` which abstracts this away entirely. The three expert implementations now have three different approaches to the same FP8 path — this is the biggest consistency risk in the PR.

Fixed. The checkpoint converter now handles the `weight` key to `gate_up_proj` so we get rid of hand-rolled fp8 control.

File: pithtrain/modules/training.py [hunk 1]
```diff
@@ -372,7 +373,9 @@ def apply_fsdp(
     )
     # FSDP recommends shard models from the bottom to the top.
     for i in range(2):
-        assert isinstance(model[i], (DeepseekV2LiteModel, GptOssModel, Qwen3MoeModel))
+        assert isinstance(
+            model[i], (DeepseekV2LiteModel, GptOssModel, Qwen3MoeModel, Qwen35MoeModel)
+        )
```
Comment (MasterJH5574) [inline pithtrain/modules/training.py]: Maybe worth to have a tuple defined outside, either globally or locally in this function, so we don't grow this for loop body every time.

Comment (haok1402) [reply pithtrain/modules/training.py]: @MasterJH5574 Fixed. It is now defined at the top of this file. I believe in the future we want to support a more convenient way to register these.

**Review by MasterJH5574 — APPROVED** [review]
